src/nnet3/nnet-simple-component.cc

```diff
@@ -295,6 +295,110 @@ void NoOpComponent::Backprop(const std::string &debug_info,
+  std::stringstream stream;
+  stream << Type() << ", dim=" << dim_
+         << ", norm-based-clipping=" << (norm_based_clipping_ ? "true" : "false")
+         << ", clipping-threshold=" << clipping_threshold_;
+  return stream.str();
+}
+
+void GradientClipperComponent::Init(int32 dim, 
+                                    BaseFloat clipping_threshold,
+                                    bool norm_based_clipping)  {
+  KALDI_ASSERT(clipping_threshold >= 0 && dim > 0);
+  dim_ = dim;
+  norm_based_clipping_ = norm_based_clipping;
+  clipping_threshold_ = clipping_threshold;
+}
+
+void GradientClipperComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 dim = 0;
+  bool ok = cfl->GetValue("dim", &dim);
+  bool norm_based_clipping = false;
+  BaseFloat clipping_threshold = 15.0;
+  cfl->GetValue("clipping-threshold", &clipping_threshold);
+  cfl->GetValue("norm-based-clipping", &norm_based_clipping);
+  if (!ok || cfl->HasUnusedValues() ||
+      clipping_threshold < 0 || dim <= 0)
+    KALDI_ERR << "Invalid initializer for layer of type "
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  Init(dim, clipping_threshold, norm_based_clipping);
+}
+
+void GradientClipperComponent::Propagate(
+                                 const ComponentPrecomputedIndexes *indexes,
+                                 const CuMatrixBase<BaseFloat> &in,
+                                 CuMatrixBase<BaseFloat> *out) const {
+  out->CopyFromMat(in);
+}
+
+void GradientClipperComponent::Backprop(const std::string &debug_info,
+                             const ComponentPrecomputedIndexes *indexes,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &out_deriv,
+                             Component *to_update, // may be NULL; may be identical
+                             // to "this" or different.
+                             CuMatrixBase<BaseFloat> *in_deriv) const {
+  // the following statement will do nothing if in_deriv and out_deriv have same
+  // memory.
+  in_deriv->CopyFromMat(out_deriv);
+  BaseFloat norm_before_clipping = 0,
+            norm_after_clipping = 0;
+
+  if (GetVerboseLevel() >= 2) {
+    norm_before_clipping = out_deriv.FrobeniusNorm();
+  }
+  if (clipping_threshold_ > 0) {
+    if (norm_based_clipping_) {
+      // each row in the derivative matrix, which corresponds to one sample in
+      // the mini-batch, is scaled to have a max-norm of clipping_threshold_
+      BaseFloat clipping_threshold_2 = pow(clipping_threshold_, 2);
+      CuVector<BaseFloat> deriv_norm(in_deriv->NumRows());
+      deriv_norm.AddDiagMat2(1.0 / in_deriv->NumCols(),
+                             *in_deriv, kNoTrans, 0.0);
+      deriv_norm.ApplyFloor(clipping_threshold_2);
+      deriv_norm.Scale(1.0 / clipping_threshold_2);
+      deriv_norm.ApplyPow(-0.5);
+      in_deriv->MulRowsVec(deriv_norm);
```

Comment:
BTW, once you move those functions to the base-class, there is a bunch of
dynamic casting in calling code (with associated sanity checks) which is no
longer necessary and can be removed.
Dan


On Mon, Aug 31, 2015 at 5:58 PM, Daniel Povey <dpovey@gmail.com> wrote:

> That is the point.  They tend to be called only in model-averaging
> operations, and you want everything to have it applied..  If someone
> decides that they want to only apply those functions for updatable
> components, they can just test the Properties() and apply it selectively.
> Dan
>
>
> On Mon, Aug 31, 2015 at 5:54 PM, Vijayaditya Peddinti <
> notifications@github.com> wrote:
>
>> In src/nnet3/nnet-simple-component.cc
>> <https://github.com/kaldi-asr/kaldi/pull/103#discussion_r38364344>:
>>
>> > +
>> > +  if (GetVerboseLevel() >= 2) {
>> > +    norm_before_clipping = out_deriv.FrobeniusNorm();
>> > +  }
>> > +  if (clipping_threshold_ > 0) {
>> > +    if (norm_based_clipping_) {
>> > +      // each row in the derivative matrix, which corresponds to one sample in
>> > +      // the mini-batch, is scaled to have a max-norm of clipping_threshold_
>> > +      BaseFloat clipping_threshold_2 = pow(clipping_threshold_, 2);
>> > +      CuVector<BaseFloat> deriv_norm(in_deriv->NumRows());
>> > +      deriv_norm.AddDiagMat2(1.0 / in_deriv->NumCols(),
>> > +                             *in_deriv, kNoTrans, 0.0);
>> > +      deriv_norm.ApplyFloor(clipping_threshold_2);
>> > +      deriv_norm.Scale(1.0 / clipping_threshold_2);
>> > +      deriv_norm.ApplyPow(-0.5);
>> > +      in_deriv->MulRowsVec(deriv_norm);
>>
>> It would also make sense to change the way the Scale() and Add() functions
>> are declared, by moving them to the Component base-class and giving them,
>> there, a default implementation that does nothing. This will allow you to
>> simplify the calling code for Scale() and Add(). You can modify your
>> GradientClipperComponent so that it overrides these functions, and uses
>> them to scale/add the stats.
>>
>> Moving the Add and Scale functions to the Component class would mean that
>> UpdateableComponent would override these methods to perform operations on
>> parameters.
>> I think it would be better if modify Add() and Scale() methods which
>> modify stats to AddStats() and ScaleStats().
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/103/files#r38364344>.
>>
>
>